### PR TITLE
feat(metadata-hash): proofs

### PR DESCRIPTION
**Pull request by josepot**

### CODE REVIEW

File: packages/merkleize-metadata/package.json [hunk 1]
```diff
@@ -42,7 +42,8 @@
     "prepack": "pnpm run build"
   },
   "dependencies": {
-    "@polkadot-api/substrate-bindings": "workspace:*"
+    "@polkadot-api/substrate-bindings": "workspace:*",
+    "@polkadot-api/utils": "workspace:*"
   },
   "devDependencies": {
     "@polkadot-api/metadata-fixtures": "workspace:*"
```
Comment (carlosala) [inline packages/merkleize-metadata/package.json:49]: Just a reminder to remove the dependency.

File: packages/merkleize-metadata/src/main.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  const lookup = getLookup(
+    definitions,
+    accessibleTypes,
+    getTypeRef,
+    getPrimitive,
+  )
+  const lookupEncoded = lookup.map(lookupType.enc)
+
+  let _proofs: Array<Uint8Array> | undefined
+  const getProofs = (): Array<Uint8Array> => {
```
Comment (carlosala) [inline packages/merkleize-metadata/src/main.ts:90]: I'd rather rename this to `getHashTree` or something similar. They are not really proofs, just the hash tree.
I confused this with `getProofData` when reviewing the code.

File: packages/merkleize-metadata/src/proof.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
+const getLevelFromIdx = (idx: number) => Math.log2(idx + 1) | 0
+const getAncestorIdx = (from: number, nLevels: number) =>
+  ((from + 1) >> nLevels) - 1
```
Comment (carlosala) [inline packages/merkleize-metadata/src/proof.ts:3]: I thought yesterday that we should add the `-1`. You were faster than me. 👏🏻

File: packages/merkleize-metadata/src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  const result = new Uint8Array(totalLen)
+
+  for (let idx = 0, at = 0; idx < len; idx++) {
+    const current = inputs[idx]
+    result.set(current, at)
+    at += current.byteLength
+  }
+
+  return result
+}
```
Comment (carlosala) [inline packages/merkleize-metadata/src/utils.ts:17]: If we are already importing `fromHex`, we can import as well `mergeUint8`.

**Review by carlosala — COMMENTED** [review]
Super duper mega job! 👏🏻

File: packages/merkleize-metadata/src/utils.ts (hunk 4 above)
Comment (josepot) [reply packages/merkleize-metadata/src/utils.ts:17]: unfortunately the `mergeUint8` that's currently on the `@polkadot-api/utils` can't be used for this... The reason being that the `inputs` come as a spread arg, and when there are too many items it crashes. It's in my TODO list to fix that. Whenever I get to it, then I will remove this one and use the fixed one.

Comment (carlosala) [reply packages/merkleize-metadata/src/utils.ts:17]: Sounds good!

**Review by carlosala — APPROVED** [review]
LET'S GOO!

**Pull request merged**